### PR TITLE
PLANNER-802: Workbench: Introduce score field delete validator

### PR DESCRIPTION
**Pull request by mcimbora**
@wmedvede This PR reworks field deletion validation from DomainHandler-based one to specialized ObjectPropertyDeleteValidator (the same approach that's used for Delete/Save/Copy/Rename validators).
Could you please take a look when you have a while? Thanks in advance!

Part of:
- https://github.com/kiegroup/optaplanner-wb/pull/168
- https://github.com/kiegroup/kie-wb-common/pull/902

### CODE REVIEW

File: kie-wb-common-screens/kie-wb-common-data-modeller/kie-wb-common-data-modeller-backend/src/main/java/org/kie/workbench/common/screens/datamodeller/backend/server/validation/ValidationServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+public class ValidationServiceImpl implements ValidationService {
+
+    private Collection<ObjectPropertyDeleteValidator> objectPropertyDeleteValidators = new ArrayList<>();
+
+    public ValidationServiceImpl() {
+    }
+
+    @Inject
+    public ValidationServiceImpl(final Instance<ObjectPropertyDeleteValidator> objectPropertyDeleteValidatorInstance) {
+        objectPropertyDeleteValidatorInstance.forEach(objectPropertyDeleteValidators::add);
```
Comment (wmedvede) [inline kie-wb-common-screens/kie-wb-common-data-modeller/kie-wb-common-data-modeller-backend/src/main/java/org/kie/workbench/common/screens/datamodeller/backend/server/validation/ValidationServiceImpl.java]: If you use 
        objectPropertyDeleteValidatorInstance.iterator().forEachRemaining(objectPropertyDeleteValidators::add);
then you'll be able to mock also the Instance 

    @Mock
    Instance<ObjectPropertyDeleteValidator> objectPropertyDeleteValidatorInstance;

and also the instance, and then you don't need this extra method void setObjectPropertyDeleteValidators(XXX)
The test will be something like this.

    @Mock
    Instance<ObjectPropertyDeleteValidator> objectPropertyDeleteValidatorInstance;

    ValidationMessage message1 = new ValidationMessage();

    ValidationMessage message2 = new ValidationMessage();


    @Before
    public void setUp() {
        ArrayList<ObjectPropertyDeleteValidator> deleteValidators = new ArrayList<>();
        ObjectPropertyDeleteValidator validator1 = (dataObject, objectProperty) -> Arrays.asList(message1);

        ValidationMessage message2 = new ValidationMessage();
        ObjectPropertyDeleteValidator validator2 = (dataObject, objectProperty) -> Arrays.asList(message2);

        deleteValidators.add(validator1);
        deleteValidators.add(validator2);

        when( objectPropertyDeleteValidatorInstance.iterator() ).thenReturn(deleteValidators.iterator());
        validationService = new ValidationServiceImpl(objectPropertyDeleteValidatorInstance);
    }

    @Test
    public void validateObjectPropertyDeletion() {

        Collection<ValidationMessage> result = validationService.validateObjectPropertyDeletion(new DataObjectImpl(),
                                                                                                new ObjectPropertyImpl());

        assertEquals(2,
                     result.size());
        assertTrue(result.contains(message1));
        assertTrue(result.contains(message2));
    }

File: kie-wb-common-screens/kie-wb-common-data-modeller/kie-wb-common-data-modeller-backend/src/main/java/org/kie/workbench/common/screens/datamodeller/backend/server/validation/ValidationServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+                                                                        final ObjectProperty objectProperty) {
+        return objectPropertyDeleteValidators
+                .stream()
+                .flatMap(c -> c.validate(dataObject,
+                                         objectProperty).stream())
+                .collect(Collectors.toList());
+    }
+
+    // Test setter
+    void setObjectPropertyDeleteValidators(final Collection<ObjectPropertyDeleteValidator> objectPropertyDeleteValidators) {
```
Comment (wmedvede) [inline kie-wb-common-screens/kie-wb-common-data-modeller/kie-wb-common-data-modeller-backend/src/main/java/org/kie/workbench/common/screens/datamodeller/backend/server/validation/ValidationServiceImpl.java]: we can probalby avoid this extra method, see the suggestion above for refactoring the test

File: kie-wb-common-screens/kie-wb-common-data-modeller/kie-wb-common-data-modeller-client/src/main/java/org/kie/workbench/common/screens/datamodeller/client/widgets/editor/DataObjectBrowserViewImpl.java [hunk 3]
```diff
@@ -339,38 +365,49 @@ public void update(int index,
                                        Style.Unit.PX);
     }
 
-    boolean isRemoveButtonEnabled(final ObjectProperty objectProperty) {
-        return domainHandlers.stream().noneMatch(h -> h.isDomainSpecificProperty(objectProperty));
+    void safeObjectPropertyDelete(final ObjectProperty objectProperty,
```
Comment (wmedvede) [inline kie-wb-common-screens/kie-wb-common-data-modeller/kie-wb-common-data-modeller-client/src/main/java/org/kie/workbench/common/screens/datamodeller/client/widgets/editor/DataObjectBrowserViewImpl.java]: Hello @mcimbora I believe the changes and the approach is good, but the invocation to the ValidatonService and this logic here should go in the Presenter

**Comment by wmedvede** [comment]
Hello @mcimbora looks good, just a couple of modifications I believe are worth.
Thanks

File: kie-wb-common-screens/kie-wb-common-data-modeller/kie-wb-common-data-modeller-backend/src/main/java/org/kie/workbench/common/screens/datamodeller/backend/server/validation/ValidationServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+import org.kie.workbench.common.services.datamodeller.core.DataObject;
+import org.kie.workbench.common.services.datamodeller.core.ObjectProperty;
+
+@Service
+@ApplicationScoped
+public class ValidationServiceImpl implements ValidationService {
+
+    private Collection<ObjectPropertyDeleteValidator> objectPropertyDeleteValidators = new ArrayList<>();
+
+    public ValidationServiceImpl() {
```
Comment (wmedvede) [inline kie-wb-common-screens/kie-wb-common-data-modeller/kie-wb-common-data-modeller-backend/src/main/java/org/kie/workbench/common/screens/datamodeller/backend/server/validation/ValidationServiceImpl.java]: in general around the platform (but no all the people do it) we add a comment on this constructors like
    //Empty constructor for Weld proxying
for remembering why we are adding this apparently not needed constructor.

**Comment by mcimbora** [comment]
@wmedvede Hi Walter, I updated the PR as suggested. Sorry for the delay, I've just returned from PTO. Would you mind checking also optaplanner-wb part of the changeset (https://github.com/kiegroup/optaplanner-wb/pull/168)? It contains planner field validator implementation and a minor fix related to score selection (allow selecting no value).

**Pull request merged**